test/support/test_api.ex

```diff
@@ -0,0 +1,15 @@
+defmodule Telepoison.TestApi do
+  use ExUnit.Case
+
+  def set_env(key, value) do
+    env_to_restore = Application.get_env(:telepoison, key, :unset)
+    Application.put_env(:telepoison, key, value)
+
+    on_exit(fn ->
+      case env_to_restore do
+        :unset -> Application.delete_env(:telepoison, key)
+        _ -> Application.put_env(:telepoison, key, env_to_restore)
+      end
+    end)
+  end
+end
```

Comment:
I think instead we could do:

```suggestion
defmodule Telepoison.TestCase do
  use ExUnit.CaseTemplate, async: false
  
  setup do
    env_to_restore = Application.get_env_all(:telepoison)

    on_exit(fn ->
        _ -> Application.put_all_env([{:telepoison, env_to_restore}])
      end
    end)
  end
end
```

and then `use Telepoison.TestCase` instead of `ExUnit.Case`

Reply:
After testing I relalized that this is broken because it doesn't remove keys that didn't exist before, only overwrites them. This should work

```suggestion
defmodule Telepoison.TestCase do
  use ExUnit.CaseTemplate, async: false

  setup do
    env_to_restore = Application.get_all_env(:telepoison)

    on_exit(fn ->
      Application.get_all_env(:telepoison)
      |> Keyword.keys()
      |> Enum.map(fn k -> Application.delete_env(:telepoison, k) end)

      Application.put_all_env([{:telepoison, env_to_restore}])
    end)
  end
end
```

Reply:
Since there are only two keys I just added two deletes for now but yes, `put_all_env` with an empty list doesn't do the job :disappointed: